File: QEasyWidgets/Components/Player.py
from typing import Optional, overload
from PySide6.QtGui import *
from PySide6.QtCore import *
from PySide6.QtWidgets import *
from PySide6.QtMultimedia import QMediaPlayer, QAudioOutput

from ..Common.Icon import *
from ..Common.StyleSheet import *
from ..Common.QFunctions import *
from .Button import ButtonBase
from .Slider import SliderBase

##############################################################################################################################

class MediaPlayerBase(QWidget):
    """
    Base class for mediaPlayer components
    """
    def __init__(self, parent: Optional[QWidget] = None):
        super().__init__(parent)

        self.StackedWidget = QStackedWidget()
        self.StackedWidget.setMaximumSize(36, 36)
        self.StackedWidget.setContentsMargins(0, 0, 0, 0)
        self.PlayButton = ButtonBase()
        self.PlayButton.setIcon(IconBase.Play)
        self.PlayButton.setBorderless(True)
        self.PlayButton.setTransparent(True)
        self.PauseButton = ButtonBase()
        self.PauseButton.setBorderless(True)
        self.PauseButton.setTransparent(True)
        self.PauseButton.setIcon(IconBase.Pause)
        self.PauseButton.clicked.connect(lambda: self.StackedWidget.setCurrentWidget(self.PlayButton))
        self.PlayButton.clicked.connect(lambda: self.StackedWidget.setCurrentWidget(self.PauseButton))
        self.StackedWidget.addWidget(self.PlayButton)
        self.StackedWidget.addWidget(self.PauseButton)
        self.StackedWidget.setCurrentWidget(self.PlayButton)

        self.Slider = SliderBase(Qt.Horizontal)

        HBoxLayout = QHBoxLayout(self)
        HBoxLayout.setSpacing(12)
        HBoxLayout.setContentsMargins(21, 12, 21, 12)
        HBoxLayout.addWidget(self.StackedWidget, stretch = 1)
        HBoxLayout.addWidget(self.Slider, stretch = 5)

        AudioOutput = QAudioOutput(self)
        self.MediaPlayer = QMediaPlayer()
        self.MediaPlayer.setAudioOutput(AudioOutput)
        #self.MediaPlayer.mediaStatusChanged.connect(lambda Status: self.MediaPlayer.stop() if Status == QMediaPlayer.EndOfMedia else None)

        StyleSheetBase.Player.apply(self)

    def setMediaPlayer(self, MediaPath: str):
        self.MediaPlayer.setSource(QUrl.fromLocalFile(MediaPath))

        self.PlayButton.clicked.connect(self.MediaPlayer.play)
        self.PauseButton.clicked.connect(self.MediaPlayer.pause)
        self.MediaPlayer.mediaStatusChanged.connect(lambda status: self.StackedWidget.setCurrentWidget(self.PlayButton) if status == QMediaPlayer.EndOfMedia else None)

        self.Slider.setRange(0, 100)
        self.Slider.sliderMoved.connect(lambda: self.MediaPlayer.setPosition(int(self.Slider.value() / 100 * self.MediaPlayer.duration())))
        self.MediaPlayer.positionChanged.connect(lambda Position: self.Slider.setValue(int(Position / self.MediaPlayer.duration() * 100)))

    def releaseMediaPlayer(self):
        self.MediaPlayer.stop()
        self.MediaPlayer.setSource('')
        #self.MediaPlayer.deleteLater()

    def setBorderless(self, borderless: bool) -> None:
        self.setProperty("isBorderless", borderless)

    def setTransparent(self, transparent: bool) -> None:
        self.setProperty("isTransparent", transparent)

    def clearDefaultStyleSheet(self) -> None:
        StyleSheetBase.Player.deregistrate(self)

##############################################################################################################################

File: c++/QEasyWidgets/Components/Player.cpp
#include "Player.h"

#include <QUrl>

#include "../Common/Icon.h"

/**
 * MediaPlayerBase implementation
 */

MediaPlayerBase::MediaPlayerBase(QWidget *parent)
    : QWidget(parent)
    , m_stackedWidget(nullptr)
    , m_playButton(nullptr)
    , m_pauseButton(nullptr)
    , m_slider(nullptr)
    , m_mediaPlayer(nullptr)
    , m_audioOutput(nullptr)
    , m_layout(nullptr) {
    init();
}

QMediaPlayer *MediaPlayerBase::mediaPlayer() const {
    return m_mediaPlayer;
}

void MediaPlayerBase::setSource(const QUrl &url) {
    if (m_mediaPlayer) {
        m_mediaPlayer->setSource(url);
    }
}

void MediaPlayerBase::play() {
    if (m_mediaPlayer) {
        m_mediaPlayer->play();
        m_stackedWidget->setCurrentWidget(m_pauseButton);
    }
}

void MediaPlayerBase::pause() {
    if (m_mediaPlayer) {
        m_mediaPlayer->pause();
        m_stackedWidget->setCurrentWidget(m_playButton);
    }
}

void MediaPlayerBase::stop() {
    if (m_mediaPlayer) {
        m_mediaPlayer->stop();
        m_stackedWidget->setCurrentWidget(m_playButton);
    }
}

void MediaPlayerBase::init() {
    setupUI();
    setupConnections();

    m_audioOutput = new QAudioOutput(this);
    m_mediaPlayer = new QMediaPlayer(this);
    m_mediaPlayer->setAudioOutput(m_audioOutput);
}

void MediaPlayerBase::setupUI() {
    m_layout = new QHBoxLayout(this);
    m_layout->setSpacing(12);
    m_layout->setContentsMargins(21, 12, 21, 12);

    // Play/Pause button stack
    m_stackedWidget = new QStackedWidget(this);
    m_stackedWidget->setMaximumSize(36, 36);
    m_stackedWidget->setContentsMargins(0, 0, 0, 0);

    m_playButton = new ButtonBase(this);
    m_playButton->setIcon(IconBase::Play);
    m_playButton->setBorderless(true);
    m_playButton->setTransparent(true);

    m_pauseButton = new ButtonBase(this);
    m_pauseButton->setIcon(IconBase::Pause);
    m_pauseButton->setBorderless(true);
    m_pauseButton->setTransparent(true);

    m_stackedWidget->addWidget(m_playButton);
    m_stackedWidget->addWidget(m_pauseButton);
    m_stackedWidget->setCurrentWidget(m_playButton);

    // Progress slider
    m_slider = new SliderBase(Qt::Horizontal, this);

    m_layout->addWidget(m_stackedWidget, 1);
    m_layout->addWidget(m_slider, 5);
}

void MediaPlayerBase::setupConnections() {
    connect(m_playButton, &QPushButton::clicked, this, &MediaPlayerBase::onPlayClicked);
    connect(m_pauseButton, &QPushButton::clicked, this, &MediaPlayerBase::onPauseClicked);

    connect(m_slider, QOverload<int>::of(&SliderBase::valueChanged),
            this, &MediaPlayerBase::onSliderValueChanged);
}

void MediaPlayerBase::onPlayClicked() {
    play();
}

void MediaPlayerBase::onPauseClicked() {
    pause();
}

void MediaPlayerBase::onPositionChanged(qint64 position) {
    if (m_mediaPlayer && m_mediaPlayer->duration() > 0) {
        int value = static_cast<int>((position * 100) / m_mediaPlayer->duration());
        m_slider->blockSignals(true);
        m_slider->setValue(value);
        m_slider->blockSignals(false);
    }
}

void MediaPlayerBase::onDurationChanged(qint64 duration) {
    Q_UNUSED(duration);
    // Duration changed, can update slider range if needed
}

void MediaPlayerBase::onSliderValueChanged(int value) {
    if (m_mediaPlayer && m_mediaPlayer->duration() > 0) {
        qint64 position = (value * m_mediaPlayer->duration()) / 100;
        m_mediaPlayer->setPosition(position);
    }
}